### PR TITLE
Implement missing_methods.rb diff in Sorbet

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This drastically reduces the time it takes to run missing_methods.rb on Stripe's codebase.

In the future, we can expand this to be a tool that will work for minimizing more kinds of RBIs, but I'm leaving that for future work.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Ke15SJq2cQa45h
→ https://go/builds/bui_Ke15gFDtwlZWTe

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Ke1n1y0aX3Q1jP
→ https://go/builds/bui_Ke1nJv2FFMVd63

File: main/minimize/minimize.cc [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+        return OutputCategory::Chalk;
+    } else if (absl::StartsWith(fullName, "PrisonGuard") || absl::StartsWith(fullName, "Critic") ||
+               absl::StartsWith(fullName, "Primus")) {
+        return OutputCategory::Util;
+    } else {
+        return OutputCategory::External;
+    }
+}
+
+core::MethodRef closestOverridenMethod(const core::GlobalState &gs, core::SymbolRef enclosingClassSymbol,
```
Comment (jvilk-stripe) [inline main/minimize/minimize.cc:71]:
```suggestion
core::MethodRef closestOverridenMethod(const core::GlobalState &gs, core::ClassOrModuleRef enclosingClassSymbol,
```

Comment (jez) [reply main/minimize/minimize.cc:71]: i'll cling to my SymbolRef's kicking and screaming until it's the only option 😅

File: main/minimize/minimize.cc [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+    auto superMethod = superClass.data(gs)->findMember(gs, name);
+    if (superMethod.exists()) {
+        return superMethod.asMethodRef();
+    } else {
+        return closestOverridenMethod(gs, superClass, name);
+    }
+}
+
+void writeClassDef(const core::GlobalState &rbiGS, options::PrinterConfig &outfile, core::SymbolRef rbiEntry,
```
Comment (jvilk-stripe) [inline main/minimize/minimize.cc:96]:
```suggestion
void writeClassDef(const core::GlobalState &rbiGS, options::PrinterConfig &outfile, core::ClassOrModuleRef rbiEntry,
```

File: definition_validator/validator.cc [hunk 3]
```diff
@@ -535,8 +535,12 @@ void validateSuperClass(core::Context ctx, const core::ClassOrModuleRef sym, con
 
     if (auto e = ctx.state.beginError(core::Loc(sym.data(ctx)->loc().file(), classDef.declLoc),
                                       core::errors::Resolver::NonClassSuperclass)) {
-        e.setHeader("The super class `{}` of `{}` does not derive from `{}`", superClass.show(ctx), sym.show(ctx),
+        auto superClassFqn = superClass.show(ctx);
+        e.setHeader("The super class `{}` of `{}` does not derive from `{}`", superClassFqn, sym.show(ctx),
                     core::Symbols::Class().show(ctx));
+        for (const auto loc : superClass.data(ctx)->locs()) {
+            e.addErrorLine(loc, "{} defined here", superClassFqn);
+        }
```
Comment (jez) [inline definition_validator/validator.cc:543]: This is technically unrelated to the change. It just helped debugging some stuff, and I figured I'd keep it.

File: main/options/options.cc [hunk 4]
```diff
@@ -17,62 +17,68 @@
...
+    {"missing-constants", &Printers::MissingConstants},
+    {"autogen", &Printers::Autogen},
+    {"autogen-msgpack", &Printers::AutogenMsgPack},
+    {"autogen-classlist", &Printers::AutogenClasslist},
     {"autogen-autoloader", &Printers::AutogenAutoloader, true, false},
-    {"autogen-subclasses", &Printers::AutogenSubclasses, true},
-    {"package-tree", &Printers::Packager},
+    {"autogen-subclasses", &Printers::AutogenSubclasses},
+    {"package-tree", &Printers::Packager, false},
+    {"minimized-rbi", &Printers::MinimizeRBI},
```
Comment (jez) [inline main/options/options.cc:81]: This `minimized-rbi` is the new option.

The rest are just because I didn't like the defaults for `PrintOptions` and wanted to flip them, which I did in `1506a85d8fca5a04e6de15c026e2a4d50e12c7be` if you want to see the minimal diff for that.

File: main/options/options.cc [hunk 5]
```diff
@@ -772,6 +788,15 @@ void readOptions(Options &opts,
...
         opts.noStdlib = raw["no-stdlib"].as<bool>();
+        opts.minimizeRBI = raw["minimize-to-rbi"].as<string>();
+        if (!opts.minimizeRBI.empty() && !opts.print.MinimizeRBI.enabled) {
+            logger->error("--minimize-to-rbi must also include --print=minimized-rbi");
+            throw EarlyReturnWithCode(1);
+        }
+        if (!opts.minimizeRBI.empty() && opts.autocorrect) {
+            logger->error("--minimize-to-rbi plus --autocorrect is not implemented");
+            throw EarlyReturnWithCode(1);
+        }
```
Comment (jez) [inline main/options/options.cc:799]: There are tests for both of these in the CLI tests.

File: main/options/options.cc [hunk 6]
```diff
@@ -402,6 +414,10 @@ buildOptions(const vector<pipeline::semantic_extension::SemanticExtensionProvide
     options.add_options("dev")("stop-after", to_string(all_stop_after),
                                cxxopts::value<string>()->default_value("inferencer"), "phase");
     options.add_options("dev")("no-stdlib", "Do not load included rbi files for stdlib");
+    options.add_options("dev")("minimize-to-rbi",
+                               "[experimental] Output a minimal RBI contining the diff between Sorbet's view of a "
+                               "codebase and the definitions present in this file",
```
Comment (jez) [inline main/options/options.cc:419]: In the future, I think it makes sense to fix the TODOs in `minimize.cc` and mark this not experimental, so that people could "minimize" a sorbet-typed RBI and a reflection generated RBI.

Right now, there are a lot of problems with it, because it aims to be a replacement for the Ruby version in Stripe's codebase, which never needed to support a lot of other features.

**Review by aprocter-stripe — APPROVED** [review]
